src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,506 @@
+
+            if (relativeUri == null)
+            {
+                return null;
+            }
+
+            var response = AwsHttpClientHelper.GetECSResponseAsync(relativeUri).GetAwaiter().GetResult();
+            var parsedReponse = BsonDocument.Parse(response);
+            var username = parsedReponse.GetValue("AccessKeyId", null)?.AsString;
+            var password = parsedReponse.GetValue("SecretAccessKey", null)?.AsString;
+            var sessionToken = parsedReponse.GetValue("Token", null)?.AsString;
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEc2Response()
+        {
+            var response = AwsHttpClientHelper.GetEC2ResponseAsync().GetAwaiter().GetResult();
+            var parsedReponse = BsonDocument.Parse(response);
+            var username = parsedReponse.GetValue("AccessKeyId", null)?.AsString;
+            var password = parsedReponse.GetValue("SecretAccessKey", null)?.AsString;
+            var sessionToken = parsedReponse.GetValue("Token", null)?.AsString;
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static string ExtractSessionTokenFromMechanismProperties(IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            string sessionToken = null;
+            if (properties != null)
+            {
+                foreach (var pair in properties)
+                {
+                    switch (pair.Key.ToUpperInvariant())
+                    {
+                        case "AWS_SESSION_TOKEN":
+                            sessionToken = pair.Value;
+                            break;
+                        default:
+                            throw new ArgumentException($"Unknown AWS property '{pair.Key}'.", nameof(properties));
+                    }
+                }
+            }
+
+            return sessionToken;
+        }
+
+        private static SecureString ToSecureString(string str)
+        {
+            if (str == null)
+            {
+                return null;
+            }
+
+            var secureString = new SecureString();
+            foreach (var c in str)
+            {
+                secureString.AppendChar(c);
+            }
+            secureString.MakeReadOnly();
+
+            return secureString;
+        }
+
+        private static void ValidateAwsCredentials(AwsCredentials awsCredentials)
+        {
+            if (awsCredentials.Username == null && (awsCredentials.Password != null || awsCredentials.SessionToken != null))
+            {
+                throw new InvalidOperationException(AccessKeyIdExceptionMessage);
```

Comment:
we usually don't use consts for that target. But I'm ok with it.

Reply:
https://github.com/mongodb/specifications/blob/afaeb1d74f1499b3eb71b9fb1eb8936868c3e5bd/source/auth/auth.rst#obtaining-credentials
not 100% sure whether these statements are equvalents:
>If a username is provided without a password (or vice-versa) or if only a session token is provided Drivers MUST raise an error

and

>In other words, regardless of how Drivers obtain credentials the only valid combination of credentials is an access key ID and a secret access key or an access key ID, a secret access key and a session token.

can the second be written in this way:

    (access key ID && a secret access key) || (an access key ID && a secret access key && a session token)

?

Reply:
if it's correct, the second statement doesn't have real value? Even if sessionToken == null from the second part, we will pass this whole condition based on the first one?
Maybe there is no problem here, but it's confusing a bit